### PR TITLE
Update versions information in HostAgent binary

### LINKED ISSUE #436: Update version information in HostAgent binary
**What steps did you take and what happened:**
```
$ ./byoh-hostagent-linux-amd64 --version
byoh-hostagent version: version.Info{Major:"1", Minor:"2", Patch:"0", BuildDate:"2022-03-16T06:59:35Z", GoVersion:"go1.16.6", Platform:"linux/amd64", Compiler:"gc"}
```
The version flag in the host agent binary displays the major, minor and patch versions. While this is good, the implementation is hard coded to be of the form `v1.2.3`. We split on `.` and then check for the length to be equal to 3 - 
https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/blob/cc9ae9bda980f2a119b494e4caf518b475151186/agent/version/version.go#L48-L67

This will break if the version is of the form `1.0.0-alpha.beta` or `1.0.0-beta.11` which are valid semantic versions. And byoh-agent displays empty fields for such verisons
```
$ ./byoh-hostagent-linux-amd64 --version
byoh-hostagent version: version.Info{Major:"", Minor:"", Patch:"", BuildDate:"2022-03-16T07:10:01Z", GoVersion:"go1.16.6", Platform:"linux/amd64", Compiler:"gc"}
```

Looking at other standard binaries like kubectl and clusterctl, they have `major` and `minor` versions only. Plus, an additional field called `GitVersion`. I think it is good practice to follow the same convention.

```
$ clusterctl version
clusterctl version: &version.Info{Major:"1", Minor:"1", GitVersion:"v1.1.2", GitCommit:"3433f7b769b4e7f5cb899b2742a5a8a1a9f51b3e", GitTreeState:"clean", BuildDate:"2022-02-17T19:14:59Z", GoVersion:"go1.17.3", Compiler:"gc", Platform:"linux/amd64"}

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:38:05Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-21T23:01:33Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}


```
**What did you expect to happen:**
byoh agent binary displays correct version information - major, minor and GitVersion info

### PR DESCRIPTION
**Pull request by sachinkumarsingh092**
**What this PR does / why we need it**:
The version flag in the host agent binary displays the major, minor and patch versions. 
This PR introduces the following flags, similar to kubectl or clusterctl, and removes the patch field from version information.
```
$ ./bin/byoh-hostagent-linux-amd64 --version
byoh-hostagent version: version.Info{Major:"0", Minor:"1", GitVersion:"v0.1.0-79-42e700c78428bb-dirty", GitCommit:"42e700c78428bb4c2096a85f5641565375d666d0", GitTreeState:"dirty", BuildDate:"2022-03-23T12:29:43Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #436 

**Additional information**
Changes:
- **.github/workflows/draft-release.yaml**: Now we extract version from the `hack/version.sh` script so no need for `AGENT_BINARY_VERSION`.
- **Makefile**: Now we use `ldflags` that are generated from the version.sh script.
- **agent/host_agent_test.go**: Tests for the agent version are changed according to the new format.
- **agent/version/version.go**: All the new required fields are introduced. We take the values of these fields from the variables generated by `ldflags` during building the binary. 
- **agent/version/version_test.go**: The tests for version.go. We build a fake binary and check its version info based on the given ldflags and compare these values. Due to the overhead of building and testing continuously, only 2 corner cases are checked which should suffice for all the cases.
- **hack/version.sh**: The script with all the logic to assign values for the different fields in the ldflags. It uses `git describe` to extract all the information required for different fields. This is taken form `sigs.k8s.io/cluster-api/hack/version.sh`. 

**Special notes for your reviewer**

### CODE REVIEW

**Comment by anusha94** [comment]
Sample Release - https://github.com/anusha94/cluster-api-provider-bringyourownhost/releases/tag/v0.0.15

**Comment by anusha94** [comment]
- Removing version_test and suite because all the logic is now moved to `hack/version.sh`
- we have one integration test in host agent to verify ldflags on the binary

File: agent/host_agent_test.go [hunk 1]
```diff
@@ -337,29 +338,47 @@ var _ = Describe("Agent", func() {
...
 			expectedStruct := version.Info{
-				Major:     "1",
-				Minor:     "2",
-				Patch:     "3",
-				BuildDate: version.BuildDate,
-				GoVersion: runtime.Version(),
-				Compiler:  runtime.Compiler,
-				Platform:  fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+				Major:        "1",
+				Minor:        "3",
```
Comment (shamsher31) [inline agent/host_agent_test.go:374]: I think the minor version is `2`, `3` is the patch version.

1 - Major
2 - Minor
3 - Patch.

Ref: https://semver.org/#summary

Comment (sachinkumarsingh092) [reply agent/host_agent_test.go]: Done. Thanks :)

File: agent/host_agent_test.go [hunk 2]
```diff
@@ -369,6 +388,69 @@ var _ = Describe("Agent", func() {
...
+
+			re = regexp.MustCompile(`Minor:\"(.*?)\"`)
+			minorOut := re.Find(out)
+			minorExpected := re.Find([]byte(expected))
+
+			re = regexp.MustCompile(`GitVersion:\"(.*?)\"`)
+			gitVersionOut := re.FindSubmatch(out)
+			gitVersionExpected := re.FindSubmatch([]byte(expected))
+
+			Expect(majorOut).Should(Equal(majorExpected))
```
Comment (anusha94) [inline agent/host_agent_test.go:448]: Can we do the same `ContainSubstring` assertion on `majorOut` and `minorOut` too? To avoid complex looking regex :)

File: agent/host_agent_test.go [hunk 3]
```diff
@@ -369,6 +388,69 @@ var _ = Describe("Agent", func() {
...
+			Expect(minorOut).Should(Equal(minorExpected))
```
Comment (anusha94) [inline agent/host_agent_test.go:449]: meta - the assertion should be closer to where its variables are defined / assigned / evaluated. That is, move the corresponding assertions closer to majorOut / minorOut blocks

File: agent/host_agent_test.go (hunk 2 above)
Comment (sachinkumarsingh092) [reply agent/host_agent_test.go:448]: The regex expressions will be the same for majorOut and minorOut. For Gitversion, we had to bypass the sanitization and so we did the FindMatch to only pull out the version info before the metadata.

File: agent/host_agent_test.go (hunk 3 above)
Comment (sachinkumarsingh092) [reply agent/host_agent_test.go:449]: Yeah looks much better that way. Will do.

File: agent/host_agent_test.go (hunk 2 above)
Comment (anusha94) [reply agent/host_agent_test.go:448]: Oh I missed gitVersion. What I am suggesting is something like this -
```
majorExpected := "Major: \"" + gitMajor + "\""
Expect(out).Should(ContainSubstring(majorExpected))
```

**Review by anusha94 — APPROVED** [review]
lgtm!

**Review by dharmjit — APPROVED** [review]
lgtm
